[C10D] C++ Callbacks part 1

wconstab opened the pull request:
Summary:
Breaking down https://github.com/pytorch/pytorch/pull/110307 into smaller
pieces to try to land without revert.

This PR adds some hook functions but does not call them.

Test Plan: OSS CI and internal tests

Differential Revision: D50460640

Skylion007 commented on torch/csrc/distributed/c10d/Hooks.cpp:
```diff
@@ -0,0 +1,37 @@
+
+namespace details {
+
+void call_collective_callbacks(EventInfo&& evt) {
```
Why does this function take as rvalue if all the callbacks are guaranteed to be const ref?

wconstab committed:
[C10D] C++ Callbacks part 0.5

Summary:
Breaking down https://github.com/pytorch/pytorch/pull/110307 into smaller
pieces to try to land without revert.

add most of the code but don't even call the event-creation logic

Test Plan: OSS CI and internal tests

Differential Revision: D50502961

wconstab committed:
[C10D] C++ Callbacks part 1 (#111582)

Summary:

Breaking down https://github.com/pytorch/pytorch/pull/110307 into smaller
pieces to try to land without revert.

This PR adds some hook functions but does not call them.

Test Plan:
OSS CI and internal tests

pyper test: https://www.internalfb.com/mlhub/pipelines/runs/fblearner/493040128

Differential Revision: D50460640
